deepreg/model/loss/util.py

```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         :return: negated loss.
         """
         return -super().call(y_true=y_true, y_pred=y_pred)
+
+
+EPS = tf.keras.backend.epsilon()
+
+
+def rectangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D rectangular
+    kernel for LocalNormalizedCrossCorrelation.
+
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
+        - kernel_vol, scalar indicating the sum of the coefficients of the equivalent
+                      3D kernel used for normalization purposes
+    """
+
+    kernel = tf.ones(shape=(kernel_size, 1, 1), dtype="float32")
+    return kernel
+
+
+def triangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D triangular
+    kernel for LocalNormalizedCrossCorrelation.
+
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
+        - kernel_vol, scalar indicating the sum of the coefficients of the equivalent
+                      3D kernel used for normalization purposes
+    """
+    fsize = int((kernel_size + 1) / 2)
+    pad_filter = tf.constant(
+        [
+            [0, 0],
+            [int((fsize - 1) / 2), int((fsize + 1) / 2)],
+            [0, 0],
+        ]
+    )
+
+    f1 = tf.ones(shape=(1, fsize, 1), dtype="float32") / fsize
+    f1 = tf.pad(f1, pad_filter, "CONSTANT")
+    f2 = tf.ones(shape=(fsize, 1, 1), dtype="float32") / fsize
+
+    kernel = tf.nn.conv1d(f1, f2, stride=[1, 1, 1], padding="SAME")
+    kernel = tf.transpose(kernel, perm=[1, 2, 0])
+
+    return kernel
+
+
+def gaussian_kernel1d_size(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D Gaussian
+    kernel for LocalNormalizedCrossCorrelation.
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
+        - kernel_vol, scalar indicating the sum of the coefficients of the equivalent
+                      3D kernel used for normalization purposes
+    """
+    mean = (kernel_size - 1) / 2.0
+    sigma = kernel_size / 3
+
+    grid = tf.range(0, kernel_size, dtype="float32")
+    grid = tf.reshape(grid, [-1, 1, 1])
+    filters = tf.exp(-tf.square(grid - mean) / (2 * sigma ** 2))
+
+    return filters
+
+
+def gaussian_kernel1d_sigma(sigma: int) -> tf.Tensor:
```

Comment:
might be a stupid question, what's the difference between the two gaussian kernels? it looks like their values are different, tho emmm what's the different utilities.

have to admit that, I didnt understand the LNCC with diff kernels, if there's any reference I can read?

Reply:
I did not find a good solution for this, sorry. For the LNCC we are specifying the size and for de multiscale loss function the sigma. I will rethink this.

About the different kernels to be fair I did not do it based on any paper, at least not in the registration literature. The LNCC basically uses low-pass filters and there are three different implementations of it. Usually (I'd say almost always) the rectangular is used, but I also tried the others on my own work and included them here as I already had the implementation. But, I'd also agree to keep things simple if you like and implement only the rectangular window.